Add support for PHP 8.2 + 8.3

larsmaedje opened the pull request.

boesing committed:
feature: introduce PHP 8.2 support

This also:
- drops support for PHP 7 and PHP 8.0
- modernizes the code, i.e. using typed properties, constructor promotion, etc.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

larsc24 committed:
php8.3 upgrade

larsc24 committed:
adjusts psr/http-message requirement

boesing commented on src/ConfigProvider.php:
```diff
@@ -4,35 +4,39 @@
 use Symfony\Component\Mime\MimeTypes;
 
+/**
+ * @psalm-import-type ServiceManagerConfigurationType from ConfigInterface
```
I am not 100% sure if I do want to have this laminas-specific dependency here.
This package does not require `laminas/laminas-servicemanager` and thus should not use anything from laminas unless its an actual dependency.

We just provide support for laminas without actually enforcing users to use laminas, therefore, we should provide the types on our own while respecting laminas specifics.

What do I mean by that? You could create a return type matching laminas requirements such as:

```php
/**
 * @psalm-type ServiceManagerConfigurationType array{factories: array<non-empty-string,callable(ContainerInterface):mixed>}
 */
```

That should provide proper return types without depending on laminas directly. For now that should already work OOTB if you use my example above as I only see inline callables.

boesing replied:
Well, after double checking, this is actually my own change I haven't merged with #3
I will take that and apply the changes accordingly.

boesing commented on .gitignore:
```diff
@@ -5,3 +5,4 @@
 /laminas-mkdoc-theme/
 /vendor/
 /psalm.xml
+.idea
```
I personally do not like having IDE specific stuff within my `.gitignore`.

You can configure your developer machine to exclude `.idea` globally by using a global `.gitignore` which matches all projects.

The rationale here is, that if every contributor adds his own IDE specific stuff to the `.gitignore`, we end up having a bunch of stuff ignored yet not related to this project (thats stuff generated from the IDE, not from tooling around this project).

Could you remove this change?

boesing commented on phpcs.xml.dist:
```diff
@@ -22,6 +22,7 @@
         <!-- We do use psalm here so the codestandard is not needed for this -->
         <exclude name="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingAnyTypeHint"/>
         <exclude name="SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingAnyTypeHint"/>
+        <exclude name="Squiz.Commenting.FunctionComment.MissingParamName"/>
```
Could you outline which line is reported so that I get an idea on why we need that here?

boesing replied:
My change as well, so nvm.

boesing commented on src/PartOfMultipartStream.php:
```diff
@@ -98,27 +87,27 @@ public function rewind(): void
     }
 
-    public function write($string)
+    public function write($string): int
```
Hm, we should have proper type-hints here:

https://github.com/php-fig/http-message/blob/402d35bcb92c70c026d1a6a9883f06b2ead23d71/src/StreamInterface.php#L115

boesing commented on src/PartOfMultipartStream.php:
```diff
@@ -98,27 +87,27 @@ public function rewind(): void
     }
 
-    public function read($length)
+    public function read($length): string
```
Hm, we should have proper type-hints here:

https://github.com/php-fig/http-message/blob/402d35bcb92c70c026d1a6a9883f06b2ead23d71/src/StreamInterface.php#L134

boesing commented on test/integration/DependenciesFromConfigProviderExtractionTrait.php:
```diff
@@ -4,24 +4,23 @@
  * @see TestCase
  *
  * @psalm-require-extends TestCase
+ * @psalm-import-type ServiceManagerConfigurationType from ConfigInterface
```
Works for me in `test` directory as we have service manager as `dev` dependency 👍🏼

boesing commented on test/integration/SymfonyMimeMimeTypeGuesserIntegrationTest.php:
```diff
@@ -4,17 +4,16 @@
 
 namespace Boesing\Psr\Http\Message\Multipart;
 
+use Boesing\Psr\Http\Message\Multipart\SymfonyMimeMimeTypeGuesser;
```
shouldn't be necessary as the namespace is already `Boesing\Psr\Http\Message\Multipart` but I guess that has been added by codesniffer? 🤷🏼‍♂️

boesing commented on test/unit/ConfigProviderTest.php:
```diff
@@ -4,6 +4,7 @@
 
 namespace Boesing\Psr\Http\Message\Multipart;
 
+use Boesing\Psr\Http\Message\Multipart\ConfigProvider;
```
shouldn't be necessary as the namespace is already `Boesing\Psr\Http\Message\Multipart` but I guess that has been added by codesniffer? 🤷🏼‍♂️

boesing commented on test/unit/ModuleTest.php:
```diff
@@ -4,12 +4,12 @@
 
 namespace Boesing\Psr\Http\Message\Multipart;
 
+use Boesing\Psr\Http\Message\Multipart\Module;
```
shouldn't be necessary as the namespace is already `Boesing\Psr\Http\Message\Multipart` but I guess that has been added by codesniffer? 🤷🏼‍♂️

boesing reviewed (changes requested):
Nice. Decent change request, thanks!
I have a few requests so that this can be merged.

I will also double check if both `psr/http-message` versions (`1.0` and `2.0`) are properly installed in CI as we should verify that we definitely support both versions.
Do we need to support both versions? Could you probably elaborate on why we have to support both? Usually, having an actual implementation of multiple major versions is tedious and I'd rather not have that here.

If there is not a strong reason why we have to support both versions, I'd drop v1.0 completely. Please let me know.

Thanks so far.

larsc24 committed:
-adds proper type hints
-removes 1.0 support for psr/http-message
-removes IDE specific gitignore

larsmaedje commented:
Regarding the support for both 1.0 and 2.0 for the psr/http-message package:
You are right, we shouldn't support both if its not necessary, so i removed that change.

boesing committed:
chore: remove `laminas/laminas-servicemanager` hard-dependency

By having the imported type in the `ConfigProvider`, that makes `laminas-servicemanager` a required dependency. To have this component decoupled from mezzio/laminas, we can introduce our own type for now.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
chore: remove support for `symfony/mime` v5 and instead add support for v7

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
chore: remove unused psalm suppression

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
Thanks, @larsmaedje!
